### ISSUE TITLE
MixUp augmentation problem

### ISSUE THREAD

**Issue by bakiuzun** [ultralytics/ultralytics#13355]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi 
While training a YOLO model with cross-validation on my custom dataset, I've implemented various augmentations (HorizontalFlip, RandomRotate90, Perspective). However, with MixUp enabled, the validation loss becomes NaN and metrics stall for some folds. Notably, the validation set uses no augmentation.It's very weird that the val become NaN only for some folds and with MixUp augmentation so I would like to know if you have any suggestion concerning this issue. 
I added the code also so you may help better 
 ```
next_img_index = (i + 1) % len(batch['img'])
next_label_1 = self._get_label_file_info(batch["im_files"][next_img_index],i)
next_img1 = batch["img"][next_img_index].cpu().numpy().transpose((1,2,0))
mixed_img1 = mixup_image(img1,next_img1)
augmented_img1 = ToTensor()(mixed_img1.astype(np.uint8))
images[i] = augmented_img1
label_1["bboxes"] = np.concatenate((label_1["bboxes"],next_label_1["bboxes"]),axis=0)
label_1["cls"] =  np.concatenate((label_1["cls"], next_label_1["cls"]), axis=0)
label_1["batch_idx"] = np.concatenate((label_1["batch_idx"],next_label_1["batch_idx"]),axis=0)   
```
here is the mixup method even though i don't think the issue come from there as for some folds everything works well
```
def mixup_image(img1,img2):
    """
    mix two image and get one image as result
    """
    alpha = 0.1
    lam =  np.random.beta(alpha,alpha)
    mixed_image1 = lam * img1 + (1 - lam) * img2
    mixed_image2 = lam * img2 + (1 - lam) * img1
    transformed_image = (mixed_image1 + mixed_image2) / 2

    return transformed_image

```

Note that _get_label_file_info method return bbox,cls and batch idx for the image path "im_files"
Im returning label_1 then for loss, metric computation.... 

Thanks

**Comment by bakiuzun** [comment]
Note: I have tried to disable amp, reduce learning rate but none of them has worked

**Comment by glenn-jocher** [comment]
Hello,

Thank you for sharing the details and the steps you've already tried. The issue with NaN losses when using MixUp augmentation could be related to the label handling during the mixing process. Here are a few suggestions:

1. **Label Clipping**: Ensure that bounding box coordinates after mixing are still valid (e.g., within image boundaries). Sometimes, mixing can lead to out-of-boundary values.

2. **Normalization Check**: Verify that the images and the mixed images are correctly normalized before being passed to the network. Improper scaling might lead to unstable training dynamics.

3. **Beta Distribution Parameter**: The `alpha` value for the Beta distribution in your MixUp method is quite low (`alpha=0.1`). This can lead to extreme values of `lam`, heavily skewing the mix towards one image or the other. Try increasing `alpha` to a value like `0.4` or `0.5` to see if it stabilizes training.

4. **Debugging Specific Folds**: Since the issue occurs only in some folds, it might be helpful to closely inspect the data in those specific folds. Check for any anomalies or differences in the data distribution or labeling.

5. **Gradual Integration**: Gradually introduce MixUp augmentation into your training regime. Start with a smaller `lam` influence and increase it as training progresses.

If these steps do not resolve the issue, consider providing more details about the specific folds that fail and any patterns or commonalities among them. This might help in pinpointing the exact cause.

**Comment by glenn-jocher** [comment]
Hello,

Thank you for the additional information. If the NaNs are originating from the `pred_distri` variable, it suggests that the issue might indeed be related to the input images or how they are being processed and fed into the model.

Here are a few steps you can take to further diagnose and potentially resolve this issue:

1. **Inspect the Mixed Images**: Before they are passed to the model, visually inspect the mixed images to ensure they are being combined correctly without any unexpected artifacts or values that could destabilize the network.

2. **Check Image Normalization**: Confirm that the normalization applied to your mixed images is consistent with the normalization expected by the network. Inconsistencies here can lead to large input values that might destabilize training.

3. **Log Values**: Temporarily add logging for the values in `pred_distri` just before the loss calculation. This might help you identify if and when they become NaN or take on extreme values.

4. **Gradual Introduction of MixUp**: If not already done, try introducing MixUp augmentation gradually into the training, starting with a very small `lam` and increasing it as training progresses to see if there is a threshold where the instability starts.

These steps should help in narrowing down the cause of the NaN values in your training process.

**Comment by bakiuzun** [comment]
Hello,  thanks for your reply. 
In my dataset I have 10% of data without label, so I made an experiment without any augmentation just by putting all the data without label at the end so the model will first see data with label and at the end of an epoch it will see data without label (I set shuffle to false).
<img width="239" alt="Screenshot 2024-06-25 at 11 01 13" src="https://github.com/ultralytics/ultralytics/assets/91346020/364ffa0a-4a33-41f1-97e8-d41f397d8888">
On the screen you can see that after the 6th epoch all 3 losses become 0 even though on the previous epochs it was not the case. I have also noticed that the cls loss start to increase when it see images without  label. 
Do you have any recommandation concerning this problem ? 
Thanks

**Comment by glenn-jocher** [comment]
Hello,

Thank you for providing additional details and the screenshot. It appears that the issue might be related to how the model handles images without labels, especially when combined with MixUp augmentation.

Here are a few recommendations to address this problem:

1. **Handling Images Without Labels**: Ensure that the model can appropriately handle images without labels. One approach is to modify the loss function to ignore these images during training. You can achieve this by setting a flag or using a mask to exclude these images from the loss calculation.

2. **MixUp Augmentation**: When using MixUp, ensure that the images without labels are not mixed with labeled images. This can be done by separating the datasets or by implementing a check within the MixUp function to skip unlabeled images.

3. **Data Shuffling**: Instead of setting `shuffle` to `False`, consider using a custom data sampler that ensures labeled images are seen first, followed by unlabeled images, but still allows for some level of randomness to avoid overfitting.

4. **Gradual Introduction of Augmentation**: Start training without any augmentation and gradually introduce MixUp after a few epochs. This can help stabilize the training process.

Here is a code snippet to modify the loss function to ignore images without labels:

```python
def custom_loss(pred, target, mask):
    # Calculate loss only for labeled images
    loss = original_loss_function(pred[mask], target[mask])
    return loss
```

And for MixUp, ensure you skip unlabeled images:

```python
def mixup_image(img1, img2, label1, label2):
    if label1 is None or label2 is None:
        return img1, label1  # Skip MixUp if any image is unlabeled
    alpha = 0.4
    lam = np.random.beta(alpha, alpha)
    mixed_image = lam * img1 + (1 - lam) * img2
    return mixed_image, lam * label1 + (1 - lam) * label2
```

Lastly, please ensure you are using the latest version of the Ultralytics packages to benefit from the latest fixes and improvements. If the issue persists, providing a minimum reproducible example would be very helpful. You can find guidance on creating one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

I hope these suggestions help! If you have any further questions or need additional assistance, feel free to ask.

**Comment by bakiuzun** [comment]
Hello,
Thank you for your help.
From what I can see, if a batch of non-labeled images is passed, the box and DFL loss are just 0, which causes the gradient to explode and the model never learn. However, even though the model's parameter gradients initially return NaN, they become numerically stable after seeing images with labels. Is this normal? This issue occurs even when using an official YOLO dataset, not just a custom one.

**Comment by glenn-jocher** [comment]
@bakiuzun hello,

Thank you for the detailed observation. It's indeed crucial to handle batches with non-labeled images properly to avoid destabilizing the training process.

To address your question:
1. **Zero Loss for Non-Labeled Images**: It's expected that the loss would be zero for non-labeled images since there are no targets to compute the loss against. However, this should not cause gradient explosions if handled correctly.
2. **Gradient Stability**: Gradients returning to stability after seeing labeled images suggests that the model can recover, but it's not ideal for training efficiency and stability.

### Recommendations:
1. **Filter Non-Labeled Images**: Ensure that batches containing non-labeled images are either skipped or handled in a way that they do not contribute to the loss calculation.
2. **Custom DataLoader**: Implement a custom DataLoader that ensures batches are composed of labeled images, especially in the initial epochs.

Here's a code snippet to filter out non-labeled images:

```python
def custom_collate_fn(batch):
    # Filter out non-labeled images
    batch = [item for item in batch if item['labels'].size > 0]
    return default_collate(batch)

# Use this custom collate function in your DataLoader
train_loader = DataLoader(dataset, batch_size=16, collate_fn=custom_collate_fn)
```

### Verify with Latest Version:
Please ensure you are using the latest version of the Ultralytics packages, as updates often include important fixes and improvements.

If the issue persists, providing a minimum reproducible example would be very helpful. You can find guidance on creating one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Thank you for your patience and cooperation. If you have any further questions, feel free to ask! 😊
